### DISCUSSION TITLE
What are the prerequisites of EJC batch compiler to produce reproducible builds?

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-jdt/eclipse.jdt.core#601] [Q&A]
Given I have the same source file and compile it multiple times, under what circumstances I could expect a [reproducible result](https://reproducible-builds.org/)?

Does it depend on

- any compiler settings except for example the `target`
- the used JDK to run the compiler or ...?
- the operating system, file encoding, architecture of the system?
- the version of EJC?
- anything else?

**Comment by iloveeclipse** [comment]
> Does it depend on
> 
> * any compiler settings except for example the `target`

Sure, something like debug attributes yes/no.

> * the used JDK to run the compiler or ...?

Yes, depending on compiler settings it might produce different bytecode if target JDKs have different (changed) API's. 

> * the operating system, file encoding, architecture of the system?

Depending on compiler settings it might inherit system/user dependent encoding for reading files. So if environment changes, produced bytecode might be different. 

> * the version of EJC?

Sure. Compiler bugs may be fixed, bytecode generated might change for various reasons (e. g. anonymous class definition order etc)

> * anything else?

Assuming system has different set of resources (free RAM, free disk space etc) during different runs, compilation with identical other parameters might produce different results (only subset of, because it could fail earlier due OOM or unable to write some classes).

Note: for all the above, I assumed classpath / modulepath dependencies do not change (physically same files).
